expfmt/encode.go

```diff
@@ -61,53 +63,113 @@ func (ec encoderCloser) Close() error {
 // FmtOpenMetrics as an option for the result. Note that this function is
 // temporary and will disappear once FmtOpenMetrics is fully supported and as
 // such may be negotiated by the normal Negotiate function.
 func NegotiateIncludingOpenMetrics(h http.Header) Format {
+	escapingScheme := Format(fmt.Sprintf("; escaping=%s", Format(model.NameEscapingScheme.String())))
 	for _, ac := range goautoneg.ParseAccept(h.Get(hdrAccept)) {
+		if escapeParam := ac.Params["escaping"]; escapeParam != "" {
+			switch Format(escapeParam) {
+			case model.EscapeNone, model.EscapeUnderscores, model.EscapeDots, model.EscapeValues:
+				escapingScheme = Format(fmt.Sprintf("; escaping=%s", escapeParam))
+			default:
+				// If the escaping parameter is unknown, ignore it.
+			}
+		}
 		ver := ac.Params["version"]
 		if ac.Type+"/"+ac.SubType == ProtoType && ac.Params["proto"] == ProtoProtocol {
+			if ac.Params["validchars"] == UTF8Valid {
+				switch ac.Params["encoding"] {
+				case "delimited":
+					return FmtProtoDelim + FmtUTF8Param
+				case "text":
+					return FmtProtoText + FmtUTF8Param
+				case "compact-text":
+					return FmtProtoCompact + FmtUTF8Param
+				}
+			}
+
 			switch ac.Params["encoding"] {
 			case "delimited":
-				return FmtProtoDelim
+				return FmtProtoDelim + escapingScheme
 			case "text":
-				return FmtProtoText
+				return FmtProtoText + escapingScheme
 			case "compact-text":
-				return FmtProtoCompact
+				return FmtProtoCompact + escapingScheme
 			}
 		}
-		if ac.Type == "text" && ac.SubType == "plain" && (ver == TextVersion || ver == "") {
-			return FmtText
+		if ac.Type == "text" && ac.SubType == "plain" && (ver == TextVersion_1_0_0 || ver == "") {
+			if ac.Params["validchars"] == UTF8Valid {
+				return FmtText_1_0_0 + FmtUTF8Param
+			}
+			return FmtText_0_0_4 + escapingScheme
+		}
+		if ac.Type == "text" && ac.SubType == "plain" && (ver == TextVersion_0_0_4 || ver == "") {
+			return FmtText_0_0_4 + escapingScheme
 		}
-		if ac.Type+"/"+ac.SubType == OpenMetricsType && (ver == OpenMetricsVersion_0_0_1 || ver == OpenMetricsVersion_1_0_0 || ver == "") {
-			if ver == OpenMetricsVersion_1_0_0 {
-				return FmtOpenMetrics_1_0_0
+		if ac.Type+"/"+ac.SubType == OpenMetricsType && (ver == OpenMetricsVersion_0_0_1 || ver == OpenMetricsVersion_1_0_0 || ver == OpenMetricsVersion_2_0_0 || ver == "") {
+			switch ver {
+			case OpenMetricsVersion_2_0_0:
+				if ac.Params["validchars"] == UTF8Valid {
+					return FmtOpenMetrics_2_0_0 + FmtUTF8Param
+				}
+				return FmtOpenMetrics_2_0_0 + escapingScheme
+			case OpenMetricsVersion_1_0_0:
+				fmt.Println("NEGOTIATE result open", FmtOpenMetrics_1_0_0+escapingScheme)
+				return FmtOpenMetrics_1_0_0 + escapingScheme
+			default:
+				return FmtOpenMetrics_0_0_1 + escapingScheme
 			}
-			return FmtOpenMetrics_0_0_1
 		}
 	}
-	return FmtText
+	return FmtText_0_0_4 + Format(fmt.Sprintf("; escaping=%s", model.EscapeValues))
```

Comment:
this specifies a default escaping scheme when none other is specified.  not sure if this is the right thing to do